classes/boardmanager.php

```diff
@@ -0,0 +1,1041 @@
+
+        $userids = $this->get_card_assignees($cardid);
+        $userids = array_unique($userids);
+
+        $update['assignees'] = $userids;
+        $update['selfassigned'] = in_array($USER->id, $userids);
+        $this->formatter->put('cards', $update);
+        $this->write_history('unassigned', constants::MOD_KANBAN_CARD, ['userid' => $userid], $card->kanban_column, $cardid);
+        helper::update_cached_timestamp($this->board->id, constants::MOD_KANBAN_CARD, $update['timemodified']);
+    }
+
+    /**
+     * Changes completion state of a card.
+     *
+     * @param int $cardid Id of the card
+     * @param int $state State
+     * @return void
+     */
+    public function set_card_complete(int $cardid, int $state) {
+        global $DB, $USER;
+        $card = $this->get_card($cardid);
+        $update = ['id' => $cardid, 'completed' => $state, 'timemodified' => time()];
+        $this->formatter->put('cards', $update);
+        $DB->update_record('kanban_card', $update);
+        $assignees = $this->get_card_assignees($cardid);
+        if ($state) {
+            helper::remove_calendar_event($this->kanban, $card, $assignees);
+        } else {
+            helper::add_or_update_calendar_event($this->kanban, $card, $assignees);
+        }
+        $card->username = fullname($USER);
+        $card->boardname = $this->kanban->name;
+        helper::send_notification($this->cminfo, 'closed', $assignees, $card, ($state == 0 ? 'reopened' : null));
+        $this->write_history(
+            ($state == 0 ? 'reopened' : 'completed'),
+            constants::MOD_KANBAN_CARD,
+            $update,
+            $card->kanban_column,
+            $cardid
+        );
+        helper::update_cached_timestamp($this->board->id, constants::MOD_KANBAN_CARD, $update['timemodified']);
+    }
+
+    /**
+     * Changes lock state of a column.
+     *
+     * @param int $columnid Id of the column
+     * @param int $state State
+     * @return void
+     */
+    public function set_column_locked(int $columnid, int $state) {
+        global $DB;
+        $update = ['id' => $columnid, 'locked' => $state, 'timemodified' => time()];
+        $DB->update_record('kanban_column', $update);
+        $this->formatter->put('columns', $update);
+        helper::update_cached_timestamp($this->board->id, constants::MOD_KANBAN_COLUMN, $update['timemodified']);
+    }
+
+    /**
+     * Changes lock state of all board columns.
+     *
+     * @param int $state State
+     * @return void
+     */
+    public function set_board_columns_locked(int $state) {
+        global $DB;
+        $columns = $DB->get_fieldset_select('kanban_column', 'id', 'kanban_board = :id', ['id' => $this->board->id]);
+        $update = ['id' => $this->board->id, 'locked' => $state, 'timemodified' => time()];
+        $DB->update_record('kanban_board', $update);
+        helper::update_cached_board($update['id']);
```

Comment:
update cached timestamp for board?

Reply:
Timestamp is not cached for board. The whole board record is cached as it is needed for every request.